src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,720 @@
+        }
+
+        size_t to_write = Min(this->current_header->xlp_rem_len, this->walPageSize - this->page_offset);
+        memcpy(((char *) this->record) + this->got_len, this->page + this->page_offset, to_write);
+        this->page_offset += MAXALIGN(to_write);
+        this->got_len += to_write;
+    }
+    this->walInterface->validXLogRecord(this->record, this->heapPageSize);
+
+    if (this->record->xl_rmid == RM_XLOG_ID && this->record->xl_info == XLOG_SWITCH)
+    {
+        this->is_switch_wal = true;
+    }
+    this->i++;
+    return true;
+}
+
+static void
+filterRecord(WalFilterState *const this)
+{
+    // In the case of overwrite contrecord, we do not need to try to filter it, since the record may not have a body at all.
+    if (this->got_len != this->record->xl_tot_len)
+    {
+        return;
+    }
+
+    const RelFileNode *node = getRelFileNodeGPDB6(this->record);
+    if (!node)
+    {
+        return;
+    }
+
+    bool isPassTheFilter = false;
+
+    // isRelationNeeded can allocate memory on the first call. Therefore, we switch the context.
+    MEM_CONTEXT_OBJ_BEGIN(this)
+    isPassTheFilter = isRelationNeeded(node->dbNode, node->spcNode, node->relNode);
+    MEM_CONTEXT_OBJ_END();
+
+    if (isPassTheFilter)
+    {
+        return;
+    }
+
+    this->record->xl_rmid = RM_XLOG_ID;
+    this->record->xl_info = XLOG_NOOP;
+    uint32_t crc = crc32cInit();
+    crc = crc32cComp(crc, (unsigned char *) XLogRecGetData(this->record), this->record->xl_len);
+    crc = crc32cComp(crc, (unsigned char *) this->record, offsetof(XLogRecord, xl_crc));
+    crc = crc32cFinish(crc);
+
+    this->record->xl_crc = crc;
+}
+
+static void
+writeRecord(WalFilterState *const this, Buffer *const output)
+{
+    uint32_t header_i = 0;
+    if (this->recPtr % this->walPageSize == 0)
+    {
+        ASSERT(!lstEmpty(this->headers));
+        const XLogPageHeaderData *const header = lstGet(this->headers, header_i);
+        const size_t to_write = XLogPageHeaderSize(header);
+
+        bufCatC(output, (const unsigned char *) header, 0, to_write);
+
+        this->recPtr += to_write;
+        header_i++;
+    }
+
+    size_t wrote = 0;
+    while (this->got_len != wrote)
+    {
+        const size_t space_on_page = this->walPageSize - bufUsed(output) % this->walPageSize;
+        const size_t to_write = Min(space_on_page, this->got_len - wrote);
+
+        bufCatC(output, (const unsigned char *) this->record, wrote, to_write);
+
+        wrote += to_write;
+
+        // write header
+        if (header_i < lstSize(this->headers))
+        {
+            bufCatC(output, lstGet(this->headers, header_i), 0, SizeOfXLogShortPHD);
+
+            this->recPtr += SizeOfXLogShortPHD;
+            header_i++;
+        }
+    }
+
+    const size_t align_size = MAXALIGN(this->got_len) - this->got_len;
+    checkOutputSize(output, align_size);
+    memset(bufRemainsPtr(output), 0, align_size);
+    bufUsedInc(output, align_size);
```

Comment:
why not changed to `bufCatC`?

Reply:
because here is the buffer part is zeroed, not copied.